CertificateSigningRequest 'signerName' field API additions

munnerz opened the pull request:
**What this PR does / why we need it**:

This implements the first phase of adding the newly proposed `signerName` field to the API. This PR *only* includes API changes:

* Adding the field to the API definitions
* Adding defaulting behaviour as described in the KEP
* Adding validation to ensure the signerName field is set
* Adding support for using `spec.signerName` as a field selector

**Special notes for your reviewer**:

In a follow up PR (within the 1.18 timeframe), we will make changes to the `approver` and `signer` components to make them signerName aware, to allow the API to be consumed more widely.

A preview of these changes (minus tests) can be seen in this changeset: https://github.com/kubernetes/kubernetes/compare/master...munnerz:csr-api-custom-signers

**Does this PR introduce a user-facing change?**:
```release-note
Add 'signerName' field to the CertificateSigningRequest API to allow for external signers to be built
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/1400
```

/kind api-change
/assign @liggitt @deads2k 
/cc @kubernetes/sig-auth-api-reviews
/sig auth

Commit:
Add signerName field to CSR resource spec

Signed-off-by: James Munnelly <james.munnelly@jetstack.io>

munnerz commented:
/priority important-soon

munnerz commented on staging/src/k8s.io/api/certificates/v1beta1/types.go:
```diff
@@ -153,3 +169,10 @@ const (
+	KubeletClientSignerName       = "kubernetes.io/kubelet-client"
+	KubeletServingSignerName      = "kubernetes.io/kubelet-serving"
+	LegacyUnknownSignerName       = "kubernetes.io/legacy-unknown"
+)
```
Not sure if these consts should also be defined in the internal API version too? Conversion treats the signerName field as an opaque string right now (i.e. we never have to translate one 'legacy' signerName into a different one), so I *think* not... what is the convention with string typed consts?

deads2k replied:
> Not sure if these consts should also be defined in the internal API version too?

external only please

munnerz commented on staging/src/k8s.io/api/certificates/v1beta1/types.go:
```diff
@@ -48,6 +48,22 @@ type CertificateSigningRequestSpec struct {
+	//  3. kubernetes.io/kubelet-serving - signs serving certificates that are honored as a valid kubelet serving certificate by the kube-apiserver, but has no other guarantees.
+	//  4. kubernetes.io/legacy-unknown - has no guarantees for trust at all.  Some distributions may honor these as client certs, but that behavior is not standard kubernetes behavior.
+	// None of these usages are related to ServiceAccount token secrets `.data[ca.crt]` in any way.
+	// You can select on this field using `.spec.signerName`.
```
I think there's some changes needed to make this *actually* possible, but I'm not too familiar with how to configure which fields should be indexed on and usable as FieldSelector's in the apiserver. If someone can provide a pointer for me, I can get this added 😄

deads2k replied:
chase this: https://github.com/kubernetes/kubernetes/blob/82123dc261819de70e02891a2bdac7d8c131f456/pkg/registry/core/secret/storage/storage.go#L53-L54

deads2k commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,73 @@ limitations under the License.
+		csr, err := ParseCSR(&certificatesv1beta1.CertificateSigningRequest{Spec: *obj})
+		if err != nil {
+			klog.Errorf("failed to parse CSR: %v", err)
+			return
```
returning early in a defaulting method worries me.  Set the value instead.

deads2k commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,73 @@ limitations under the License.
+		obj.SignerName = certificatesv1beta1.LegacyUnknownSignerName
+		csr, err := ParseCSR(&certificatesv1beta1.CertificateSigningRequest{Spec: *obj})
+		if err != nil {
+			klog.Errorf("failed to parse CSR: %v", err)
```
utilruntime.HandleError

deads2k commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,73 @@ limitations under the License.
+		case isKubeletClientCSR(csr, obj.Usages):
+			obj.SignerName = certificatesv1beta1.KubeletClientSignerName
+			break
+		}
```
> Based on my reading of the KEP, there is no defaulting behaviour for other in-built signerName - if we do want to do add them in future, we can do so here with a similar pattern

I don't anticipate any future defaulting.  This structure is ok though.

mikedanese replied:
Should we continue to sign kubelet certs with the legacy signer until they upgrade? I don't think making this defaulting logic so fancy buys us anything.

munnerz replied:
Right now we don't have a real distinction between these two signer types, and this PR does not add support for the new signer type to kube-controller-manager (nor does it add any awareness to the existing approver & signer implementation of the `signerName` field). I don't think we've decided what the real distinction between these two `signerNames` will look like in kube-controller-manager.

I'm not sure what the implications of changing the defaulting behaviour for this field in future are though if we decide to remove defaulting the `kubelet-client` now. Would this be an acceptable change in the move from v1beta1 to v1?

deads2k replied:
> Should we continue to sign kubelet certs with the legacy signer until they upgrade? I don't think making this defaulting logic so fancy buys us anything.

It means that a skewed kube-controller-manager is assured of doing the right thing with kubelet-client certificates on the upgrades.  It allows the legacy signer to be disconnected without upgrading kubelets immediately.  I think it's a worthwhile defaulting.

mikedanese replied:
We can't change the current behavior until we remove the v1beta1 API so this would be opt-in. On the other hand, we will have to carry this until we remove the v1beta1 API. I would not expect the capability that this enables to be widely used or useful, so I lean towards keeping it simple. We get to where we want to be in 2 releases and in the meantime, inclined operators can just upgrade their nodes. I don't feel terribly strongly about this, but I'm dubious that adding this defaulting will make a difference for anyone.

deads2k replied:
> We can't change the current behavior until we remove the v1beta1 API so this would be opt-in.

With defaulting, given a paired kube-apiserver and kube-controller-manager, by n+2 the kube-contoller-manager never has to consider an empty signerName ever again.  Without defaulting, empty signerNames must be tolerated until we actually add a v1.

munnerz replied:
As discussed on the call, we're going to keep this defaulting logic and require those implementing out-of-tree approval plugins to either:

1) create the appropriate RBAC role/rolebinding to allow them to continue approving CSRs
2) disable the CertificateApproval admission plugin

deads2k commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,73 @@ limitations under the License.
+		return false
+	}
+
+	if !strings.HasPrefix(req.Subject.CommonName, "system:node:") {
```
They can never change.  I don't really care about a field.  You did right by not picking up a package dep.

mikedanese commented on pkg/apis/certificates/types.go:
```diff
@@ -42,6 +42,22 @@ type CertificateSigningRequestSpec struct {
 	// Base64-encoded PKCS#10 CSR data
 	Request []byte
 
+	// Requested signer for the request.  It is a qualified name in the form: `scope-hostname.io/name`.
```
Please linewrap comments at 80 characters like the rest of this file. There's a lot of double spaces in this comment.

mikedanese commented on pkg/apis/certificates/types.go:
```diff
@@ -42,6 +42,22 @@ type CertificateSigningRequestSpec struct {
+	//  2. Otherwise, it is assigned "kubernetes.io/legacy-unknown".
+	// In v1 it will be required.  Distribution of trust for signers happens out of band.
+	// that will be honored by the kube-controller-manager signer after approval.
+	//  1. kubernetes.io/kube-apiserver-client - signs certificates that will be honored as client-certs by the kube-apiserver. Never auto-approved by kube-controller-manager.
```
Document these on the signer name const and remove this doc from here.

mikedanese commented on staging/src/k8s.io/api/certificates/v1beta1/types.go:
```diff
@@ -48,6 +48,20 @@ type CertificateSigningRequestSpec struct {
+	//     `"key encipherment", "digital signature", "client auth"`
+	//  2. Otherwise, it is assigned "kubernetes.io/legacy-unknown".
+	// In v1 it will be required.  Distribution of trust for signers happens
+	// out of band.
```
Only document v1beta1 behavior in the v1beta1 API.

munnerz commented:
As part of this PR I had to modify the `after_roundtrip` test fixtures for v1.16 and v1.17 *protobuf* formats.

The diff:

```
diff -u \
  <(tail -c +5 certificates.k8s.io.v1beta1.CertificateSigningRequest.pb | protoc --decode_raw) \
  <(tail -c +5 certificates.k8s.io.v1beta1.CertificateSigningRequest.after_roundtrip.pb | protoc --decode_raw)
--- /dev/fd/11	2020-01-06 14:25:44.000000000 +0000
+++ /dev/fd/12	2020-01-06 14:25:44.000000000 +0000
@@ -54,6 +54,7 @@
         1: "23"
       }
     }
+    7: ""
   }
   3 {
     1 {
```

From what I understand, this is expected and acceptable. But the README specifically says these changes should come under additional scrutiny, so I want to call it out here so it's not unnoticed 😄

liggitt commented:
> From what I understand, this is expected and acceptable.

Making the field a `*string` and `omitempty` will avoid serializing empty values and remove the need to modify the round trip fixtures. In v1, when the field becomes required, we could make it a `string`

munnerz commented:
/retest

munnerz commented on pkg/registry/certificates/certificates/storage/storage.go:
```diff
@@ -50,7 +52,11 @@ func NewREST(optsGetter generic.RESTOptionsGetter) (*REST, *StatusREST, *Approva
+	options := &generic.StoreOptions{
+		RESTOptions: optsGetter,
+		AttrFunc:    csrregistry.GetAttrs,
+		TriggerFunc: map[string]storage.IndexerFunc{"spec.signerName": csrregistry.SignerNameTriggerFunc},
```
After spending quite some time digging, I couldn't really work out if setting the TriggerFunc here is actually required... there are no doc comments on the field that I can see, nor can I see much explanation of what a TriggerFunc is used for.

I get the feeling this *may not* be required, as things like Pods support field selectors on many different fields (i.e. restartPolicy, nodeName, etc) yet only list a TriggerFunc for `spec.nodeName` (only a single element in this map is actually supported by the watch cache).

Does anyone know what the TriggerFunc is for? Does setting this here actually change anything? Should this particular change be reverted?

liggitt replied:
it is not required. it causes the watch cache to maintain an index for this field at https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L319-L333

We typically only use it for high cardinality resources where we expect lots of filtered watchers, like nodes watching pods

munnerz replied:
Removed

munnerz commented on pkg/registry/certificates/certificates/strategy.go:
```diff
@@ -190,3 +194,48 @@ func (csrApprovalStrategy) PrepareForUpdate(ctx context.Context, obj, old runtim
+		Label:       label,
+		Field:       field,
+		GetAttrs:    GetAttrs,
+		IndexFields: []string{"spec.signerName"},
```
This has also been cargo-culted from the `pod` package too: https://github.com/kubernetes/kubernetes/blob/dd5272b76f07bea60628af0bb793f3cca385bf5e/pkg/registry/core/pod/strategy.go#L204

Again, I'm uncertain if it is required (or whether setting the PredicateFunc is even required) given that Pods support field selectors on many different fields, yet only lists `spec.nodeName` in the IndexFields list (linked above).

Can anyone shed any light as to whether this is required, or what the effect of setting it is?

munnerz commented on pkg/registry/certificates/certificates/strategy.go:
```diff
@@ -190,3 +194,48 @@ func (csrApprovalStrategy) PrepareForUpdate(ctx context.Context, obj, old runtim
+}
+
+// NameTriggerFunc returns value metadata.namespace of given object.
+func NameTriggerFunc(obj runtime.Object) string {
```
This is not actually used, although other similar packages define (and *some* use) a similarly defined function

munnerz replied:
Removed as it is not required

munnerz commented on pkg/registry/certificates/certificates/strategy.go:
```diff
@@ -190,3 +194,48 @@ func (csrApprovalStrategy) PrepareForUpdate(ctx context.Context, obj, old runtim
+func SelectableFields(obj *certificates.CertificateSigningRequest) fields.Set {
+	objectMetaFieldsSet := generic.ObjectMetaFieldsSet(&obj.ObjectMeta, true)
+	csrSpecificFieldsSet := fields.Set{
+		"spec.signerName": valueOrEmpty(obj.Spec.SignerName),
```
If the pointer is nil, we set this to "" (the empty string value). I am fairly certain this is correct.. 😄

deads2k replied:
Yeah, that looks reasonable.

munnerz commented:
I've addressed review comments and also added an integration test for the field selector parts (i.e. validating that a `spec.signerName` field selector works).

Barring the discussion around defaulting and whether to automatically default existing `kubelet-client` requests (and further, how we handle upgrades of existing out-of-tree approval plugins) I think this is ready to merge unless anyone else has comments.

Hopefully we can resolve the above on today's sig-auth call 😄 

I've also opened #86933 which adds the admission plugin described in the KEP - I will create another PR today/tomorrow that implements the required controller-manager changes too (to the 'approver' and the 'signer' controllers).

munnerz commented:
/retest

BenTheElder commented:
/retest

munnerz commented:
Needs updating with the latest changes in https://github.com/kubernetes/enhancements/pull/1400. I'll get this done later today 😄 

/hold

munnerz commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -60,6 +62,24 @@ func ValidateCertificateSigningRequest(csr *certificates.CertificateSigningReque
+				}
+			}
+
+			// TODO: should we also validate the 'path' component?
```
@deads2k @liggitt I was uncertain exactly what validation to perform on the value of `signerName`, but these seemed to make sense. Any suggestions?

deads2k replied:
> @deads2k @liggitt I was uncertain exactly what validation to perform on the value of `signerName`, but these seemed to make sense. Any suggestions?

I favor strictness. There should be something.  I'll throw out. 
 1. length for sure. Needs to be at least one character
 2. I think I'd like to split on dots and verify each bit between the dots was at least one character and `NameIsDNSLabel`

munnerz replied:
I've updated this to check that both the domain and path component have a length of at least 1, and each component split on `.` are DNSLabels (as per `NameIsDNSLabel)`.

Let me know if there are any further restrictions you can think of here. There is no fixed max length per component, but there is a maximum overall length of 571 characters.

munnerz commented:
Updated to account for the final state of the KEP in https://github.com/kubernetes/enhancements/pull/1400.

This should now implement the initial set of changes described in that PR. Still remaining:

* Adding `sign` and `approve` verb admission controllers (https://github.com/kubernetes/kubernetes/pull/86933)
* kube-controller-manager changes for the built-in approver & signer

The remaining pieces will be completed in time for 1.18 codefreeze 😄

/hold cancel

munnerz commented:
/retest

munnerz commented:
/retest

deads2k commented on pkg/apis/certificates/fuzzer/fuzzer.go:
```diff
@@ -29,6 +29,8 @@ var Funcs = func(codecs runtimeserializer.CodecFactory) []interface{} {
 		func(obj *certificates.CertificateSigningRequestSpec, c fuzz.Continue) {
 			c.FuzzNoCustom(obj) // fuzz self without calling this function again
 			obj.Usages = []certificates.KeyUsage{certificates.UsageKeyEncipherment}
+			signerName := "custom-sample-signer"
```
make it valid at least.

deads2k commented on pkg/apis/certificates/types.go:
```diff
@@ -42,6 +42,20 @@ type CertificateSigningRequestSpec struct {
+	// Requested signer for the request. It is a qualified name in the form:
+	// `scope-hostname.io/name`.
+	// If empty, it will be defaulted for v1beta1:
+	//  1. If it's a kubelet client certificate, it is assigned
```
we also now default values for kubelet serving.  I think we can shorten this by removing the information about how we make the determination.

deads2k commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,102 @@ limitations under the License.
+	if !equalUnsorted(requiredUsages, usages) {
+		return false
+	}
+
```
missing restriction on dns names, emailaddresses, and ip addresses.

Comparing against https://github.com/kubernetes/kubernetes/blob/release-1.10/pkg/controller/certificates/approver/sarapprove.go#L179-L193

deads2k commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -58,6 +60,52 @@ func ValidateCertificateSigningRequest(csr *certificates.CertificateSigningReque
+	if csr.Spec.SignerName == nil {
+		allErrs = append(allErrs, field.Required(specPath.Child("signerName"), "signerName must be provided"))
+	} else {
+		// ensure signerName is of the form domain.com/something and up to 571 characters
```
@liggitt I remember you had a particular format in mind that you wanted to support. This is a good spot to document it.

deads2k replied:
`<dnsSubdomain>/<dnsLabel>.<dnsSubdomain>` I think

deads2k commented on staging/src/k8s.io/api/testdata/HEAD/certificates.k8s.io.v1beta1.CertificateSigningRequest.json:
```diff
@@ -42,26 +42,27 @@
   },
   "spec": {
     "request": "OA==",
+    "signerName": "19",
```
@liggitt I don't remember how the test is written. Are you expecting to see these change?

munnerz replied:
I _think_ these are expected as the `testdata` subdirectory is named `HEAD`, so I think it should reflect the latest version of the code?

liggitt replied:
yeah, changes to API types are expected to change test data in HEAD, but not in round-tripping of data snapshots from past releases

deads2k commented:
minor comments

/approve

dims commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -58,6 +60,52 @@ func ValidateCertificateSigningRequest(csr *certificates.CertificateSigningReque
+			}
+
+			// check the domain portion to ensure it consists of valid dns labels
+			domainOnDots := strings.Split(domain, ".")
```
Do we need to ensure there is at least one `.`?

munnerz replied:
I don't think that needs to be a hard requirement - I'd personally say `example/something` is a valid `signerName`, but I'm not sure what the standard in other similar fields is.

mikedanese replied:
+1 on using IsFullyQualifiedName or IsQualifiedname.

munnerz replied:
I've opted for `IsFullyQualifiedDomainName` for now as we can always loosen this requirement in future if there's a need for it. I also opted to *not* used `IsFullyQualifiedName`as that method is deprecated (and actually requires *three* `.` to be valid) 😄

munnerz commented:
@deads2k addressed your feedback https://github.com/kubernetes/kubernetes/compare/4aa8fd810289b73a0fcd94b568bf4815a1423e4c..6d1f3c87a214438957eff9c1a9ced0899207df98

munnerz commented:
/retest

munnerz commented:
Should we also validate that the `signerName` field is never updated? Seems sensible to me as the semantics of that change are hard to define.

munnerz commented:
> Should we also validate that the signerName field is never updated? Seems sensible to me as the semantics of that change are hard to define.

Ah, I see that in the `registry` we already disallow such changes:

```go
	newCSR.Spec = oldCSR.Spec
	newCSR.Status = oldCSR.Status
```

enj commented on test/integration/certificates/field_selector_test.go:
```diff
@@ -0,0 +1,107 @@
+		t.Errorf("unable to create test csr: %v", err)
+		return
+	}
+	defer client.CertificatesV1beta1().CertificateSigningRequests().Delete(csr1.Name, nil)
```
Each integration test is isolated so we should not need these `defer`s right?

enj commented on test/integration/certificates/field_selector_test.go:
```diff
@@ -0,0 +1,107 @@
+	csr1 := buildTestingCSR("csr-1", "example.com/signer-name-1")
+	csr2 := buildTestingCSR("csr-2", "example.com/signer-name-2")
+
+	if _, err := client.CertificatesV1beta1().CertificateSigningRequests().Create(csr1); err != nil {
```
nit: create a helper for creating a CSR and `t.Fatal` on any error.

enj commented on test/integration/certificates/field_selector_test.go:
```diff
@@ -0,0 +1,107 @@
+	client := clientset.NewForConfigOrDie(&restclient.Config{Host: s.URL, ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}}})
+
+	csr1 := buildTestingCSR("csr-1", "example.com/signer-name-1")
+	csr2 := buildTestingCSR("csr-2", "example.com/signer-name-2")
```
Can you add `csr3 := buildTestingCSR("csr-3", "example.com/signer-name-2")` so that we can confirm that multiple CSRs with the same signer are returned together?

enj reviewed (changes requested):
I think this is pretty close.  Thanks for working on this @munnerz

enj commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,106 @@ limitations under the License.
 		obj.Usages = []certificatesv1beta1.KeyUsage{certificatesv1beta1.UsageDigitalSignature, certificatesv1beta1.UsageKeyEncipherment}
 	}
+
+	if obj.SignerName == nil {
```
Please add an integration test for defaulting.

munnerz replied:
Is this best as an integration test or a unit test?

munnerz replied:
After discussion on Slack, I've added a basic integration test that checks everything is wired up properly, and then unit tests that covers various different cases in-depth.

enj commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -58,6 +60,54 @@ func ValidateCertificateSigningRequest(csr *certificates.CertificateSigningReque
 	if len(csr.Spec.Usages) == 0 {
 		allErrs = append(allErrs, field.Required(specPath.Child("usages"), "usages must be provided"))
 	}
+	if csr.Spec.SignerName == nil {
```
Validation probably needs unit tests.

mikedanese commented on pkg/apis/certificates/types.go:
```diff
@@ -42,6 +42,19 @@ type CertificateSigningRequestSpec struct {
 
+	// Requested signer for the request. It is a qualified name in the form:
+	// `scope-hostname.io/name`.
+	// If empty, it will be defaulted for v1beta1:
```
Can we only document this in v1beta1/types.go? Defaulting is specific to API versions.

mikedanese commented on staging/src/k8s.io/api/certificates/v1beta1/types.go:
```diff
@@ -48,6 +48,19 @@ type CertificateSigningRequestSpec struct {
 
+	// Requested signer for the request. It is a qualified name in the form:
+	// `scope-hostname.io/name`.
+	// If empty, it will be defaulted for v1beta1:
```
No need to qualify this with v1beta1. This is v1beta1 documentation.

munnerz commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -58,6 +60,54 @@ func ValidateCertificateSigningRequest(csr *certificates.CertificateSigningReque
+		// ensure signerName is of the form domain.com/something and up to 571 characters.
+		// This length and format is specified to accommodate signerNames like:
+		// <resource-type-name>.<fully-qualified-group-name>/<resource-namespace>.<resource-name>
+		if len(*csr.Spec.SignerName) > 571 {
```
Thanks for forcing this one.. because after I started adding things up, I got to 635 🙃 

```
		// ensure signerName is of the form domain.com/something and up to 635 characters.
		// This length and format is specified to accommodate signerNames like:
		// <resource-type-name>.<fully-qualified-group-name>/<resource-namespace>.<resource-name>.
		// The max length of a resource type name is 63 characters (DNS1035Label max length)
		// The max length of a group name is 253 characters (DNS1123Subdomain max length)
		// The max length of a namespace name is 63 characters (DNS1123Label max length)
		// The max length of a resource name is 253 characters (DNS1123Subdomain max length)
		// We then add an additional 3 characters to account for the two '.' and one '/'.
```

Does that make sense?

munnerz replied:
cc @liggitt as you initially came up with 571 and this is what is in the KEP

munnerz commented:
I've also updated the `kubectl describe` output for CSRs to include the signerName if it is non-nil - the explicit check for `nil` is to handle the case where a user with a newer kubectl is talking to an older apiserver: https://github.com/kubernetes/kubernetes/pull/86476/commits/a407143f0ac1c77e5a67e7d0ef18e830cdbc5e4e

enj commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -50,59 +52,53 @@ func ValidateCertificateRequestName(name string, prefix bool) []string {
+		signerNameMaxLength := 635
+		if len(csr.Spec.SignerName) > signerNameMaxLength {
+			allErrs = append(allErrs, field.TooLong(specPath.Child("signerName"), csr.Spec.SignerName, signerNameMaxLength))
 		} else {
```
This else is not necessary.  We want to provide as many invalid errors as possible.

enj commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -50,59 +52,53 @@ func ValidateCertificateRequestName(name string, prefix bool) []string {
+		if len(csr.Spec.SignerName) > signerNameMaxLength {
+			allErrs = append(allErrs, field.TooLong(specPath.Child("signerName"), csr.Spec.SignerName, signerNameMaxLength))
 		} else {
-			domain := parts[0]
```
Can we break the signer name validation into its own function (with early returns).  The `if/else` blocks are hard to keep track of.

enj commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -50,59 +52,53 @@ func ValidateCertificateRequestName(name string, prefix bool) []string {
+					// check the path portion to ensure it consists of valid dns labels
+					pathOnDots := strings.Split(path, ".")
+					for _, s := range pathOnDots {
+						if errs := validation.NameIsDNSLabel(s, false); len(errs) > 0 {
```
You do not need this `len` check.

`for _, err := range validation.NameIsDNSLabel(s, false) {`

enj commented on test/integration/certificates/field_selector_test.go:
```diff
@@ -23,7 +23,9 @@ import (
 	clientset "k8s.io/client-go/kubernetes"
+	certclientset "k8s.io/client-go/kubernetes/typed/certificates/v1beta1"
 	restclient "k8s.io/client-go/rest"
+
```
nit: random space?

enj commented on pkg/apis/certificates/validation/validation_test.go:
```diff
@@ -0,0 +1,237 @@
+	}
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			el := ValidateCertificateSigningRequest(&test.csr)
```
See earlier comment, but we def need some update tests as well.

munnerz replied:
What should those update tests cover? We don't perform validation on Update as the registry already nils out changes

enj replied:
I think we can ignore this per the defaulting on read thread.

enj commented on pkg/apis/certificates/validation/validation_test.go:
```diff
@@ -0,0 +1,237 @@
+					SignerName: validSignerName,
+				},
+			},
+			errs: field.ErrorList{
```
While I think this is fine, I kind of like to compare the `string` output of the errors because that is the actual error message the user will see.  i.e. if you do a `struct` level comparison, you will not notice if your error message dumps out like a memory address or something similar.

enj commented on pkg/apis/certificates/validation/validation_test.go:
```diff
@@ -0,0 +1,237 @@
+				field.Invalid(specPath.Child("signerName"), "example.com/-", "a DNS-1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')"),
+			},
+		},
+		"signerName path component contains no '.' but is valid": {
```
Add a test case which has dots in the path?

enj commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -50,59 +52,53 @@ func ValidateCertificateRequestName(name string, prefix bool) []string {
+				allErrs = append(allErrs, utilvalidation.IsFullyQualifiedDomainName(specPath.Child("signerName"), domain)...)
+
+				// ensure the path portion has a length greater than 0
+				if len(path) == 0 {
```
We should probably limit the path length to something sensible too.

munnerz replied:
Updated after discussion on Slack https://kubernetes.slack.com/archives/C0EN96KUY/p1581090157035800

enj reviewed (changes requested):
I still need to review 9ad5f140abc140d3fbb2431577c4f5974e9c937e

munnerz commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -21,7 +21,9 @@ import (
 	"k8s.io/apimachinery/pkg/api/validation"
+	utilvalidation "k8s.io/apimachinery/pkg/util/validation"
 	"k8s.io/apimachinery/pkg/util/validation/field"
+
```
This space is added by my IDE 😬 it sees the staging repos as external repositories

munnerz commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -50,59 +52,53 @@ func ValidateCertificateRequestName(name string, prefix bool) []string {
 	}
-	if csr.Spec.SignerName == nil {
+	if csr.Spec.SignerName == "" {
 		allErrs = append(allErrs, field.Required(specPath.Child("signerName"), "signerName must be provided"))
```
Defaults are applied on reads, so if an apiserver is running this version of validation, it must be running the defaulting logic, so this field will always be set.

The only case like this to handle is *client side* if we want to ensure that a newer *client* can talk to an older *apiserver* (i.e. the changes in controller-manager and kubectl). Officially I don't think this is supported, but practically it's good to do. Regardless, this shouldn't be an issue here as the apiserver is up to date.

> Since we do not allow .spec changes on update via the strategy, ValidateCertificateSigningRequestUpdate should be updated to not validate .spec at all.

```
func ValidateCertificateSigningRequestUpdate(newCSR, oldCSR *certificates.CertificateSigningRequest) field.ErrorList {
	validationErrorList := ValidateCertificateSigningRequest(newCSR)
	metaUpdateErrorList := apivalidation.ValidateObjectMetaUpdate(&newCSR.ObjectMeta, &oldCSR.ObjectMeta, field.NewPath("metadata"))
	return append(validationErrorList, metaUpdateErrorList...)
}
```

Is the current method - do you mean we should remove calling ValidateCertificateSigningRequest altogether? This isn't code added by me, and it's standard throughout all other validators so it seemed sensible to keep it there 😄

munnerz replied:
Unless you mean in `csrStatusStrategy` we call validate update - in the case we are still fine, because when the object is decoded it'll be defaulted before the validation functions are called, meaning signerName will still be set even if a client omits it.

enj replied:
> Defaults are applied on reads, so if an apiserver is running this version of validation, it must be running the defaulting logic, so this field will always be set.

For some reason I thought we only defaulted on writes, but yeah, if we default on reads then this is a non-issue.

> Is the current method - do you mean we should remove calling ValidateCertificateSigningRequest altogether? This isn't code added by me, and it's standard throughout all other validators so it seemed sensible to keep it there 😄

We do not need to change it since it causes no issue.  I was mostly saying that validating `.spec` does not make sense since we guarantee `.spec` does not change - i.e. we have asserted that it is and will always be valid on the initial `create`.

enj commented:
> 9ad5f14

9ad5f140abc140d3fbb2431577c4f5974e9c937e LGTM

liggitt commented on pkg/apis/certificates/types.go:
```diff
@@ -42,6 +42,12 @@ type CertificateSigningRequestSpec struct {
+	// Requested signer for the request. It is a qualified name in the form:
+	// `scope-hostname.io/name`.
+	// Distribution of trust for signers happens out of band.
+	// You can select on this field using `.spec.signerName`.
```
the field selector is `spec.signerName`, no leading dot

liggitt commented on staging/src/k8s.io/api/certificates/v1beta1/types.go:
```diff
@@ -48,6 +48,19 @@ type CertificateSigningRequestSpec struct {
+	//     "kubernetes.io/kubelet-serving".
+	//  3. Otherwise, it is assigned "kubernetes.io/legacy-unknown".
+	// Distribution of trust for signers happens out of band.
+	// You can select on this field using `.spec.signerName`.
```
```suggestion
	// You can select on this field using `spec.signerName`.
```

liggitt commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -48,19 +51,96 @@ func ValidateCertificateRequestName(name string, prefix bool) []string {
+		// return early here as we should not continue attempting to validate a missing or malformed path segment
+		// (i.e. one containing multiple or zero `/`)
+		return el
+	}
```
suggest moving this right after `segments := strings.Split(signerName, "/")`, the returned example in the message is sufficient

liggitt commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,121 @@ limitations under the License.
+	if err != nil {
+		// This case shouldn't be reached as we already have validation in
+		// admission that the 'request' data is a valid CSR.
+		utilruntime.HandleError(err)
```
don't log here... incoming API requests can contain invalid CSR data...

liggitt commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,121 @@ limitations under the License.
+	case isKubeletServingCSR(obj, csr):
+		return certificatesv1beta1.KubeletServingSignerName
+	}
+	return certificatesv1beta1.LegacyUnknownSignerName
```
nit:
```go
switch {
case err != nil:
	return certificatesv1beta1.LegacyUnknownSignerName
case isKubeletClientCSR(obj, csr):
	 return certificatesv1beta1.KubeAPIServerClientKubeletSignerName
case isKubeletServingCSR(obj, csr):
	 return certificatesv1beta1.KubeletServingSignerName
default:
	 return certificatesv1beta1.LegacyUnknownSignerName
}
```

liggitt commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,121 @@ limitations under the License.
+
+	if spec.Username != req.Subject.CommonName {
+		return false
+	}
```
```
if spec.Username != req.Subject.CommonName {
	return false
}
```

I didn't expect this... who submits the CSR shouldn't impact whether it is considered kubelet serving CSR

enj replied:
I thought this was supposed to be the "self-rotate" flow.

munnerz replied:
This was based on the code pulled from release-1.11 - so 'cargo culted' as it were. Happy to remove if it should be removed 😄

liggitt replied:
that code was checking 1) was it a kubelet serving cert and 2) was the caller authorized

here we only care about 1

liggitt commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,121 @@ limitations under the License.
+
+// defaultSignerNameFromSpec will determine the signerName that should be set
+// by attempting to inspect the 'request' content and the spec options.
+func defaultSignerNameFromSpec(obj *certificatesv1beta1.CertificateSigningRequestSpec) string {
```
from the KEP:

> To support HA upgrades, the kube-controller-manager will duplicate defaulting code for an empty signerName for one release.

that means this needs to be exported, right?

liggitt commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,121 @@ limitations under the License.
+	return true
+}
+
+func isKubeletClientCSR(spec *certificatesv1beta1.CertificateSigningRequestSpec, req *x509.CertificateRequest) bool {
```
I would expect kube-controller-manager to use this same logic to determine whether it should approve and sign a certificate that requests the apiserver client signerName. You should only need `req *x509.CertificateRequest` as a parameter, and I would suggest moving it to `pkg/apis/certificates/helpers.go` and exporting it

munnerz replied:
It also needs to accept a list of usages, which requires passing a versioned var - but I could keep it in this package and change the function to accept `(req *x509.CertificateRequest, usages []capi.KeyUsage)` then export it?

liggitt replied:
ah, right... usages

liggitt commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,121 @@ limitations under the License.
+	return certificatesv1beta1.LegacyUnknownSignerName
+}
+
+func isKubeletServingCSR(spec *certificatesv1beta1.CertificateSigningRequestSpec, req *x509.CertificateRequest) bool {
```
I would expect kube-controller-manager to use this same logic to determine whether it should approve and sign a certificate that requests the kubelet serving signerName. You should only need `req *x509.CertificateRequest` as a parameter, and I would suggest moving it to `pkg/apis/certificates/helpers.go` and exporting it

munnerz replied:
Again, requires usages.. I will export it but keep it in this package - this code isn't 'staged' anyway so its location shouldn't make too much of a difference?

liggitt commented on pkg/apis/certificates/v1beta1/defaults.go:
```diff
@@ -17,15 +17,121 @@ limitations under the License.
+// defaultSignerNameFromSpec will determine the signerName that should be set
+// by attempting to inspect the 'request' content and the spec options.
+func defaultSignerNameFromSpec(obj *certificatesv1beta1.CertificateSigningRequestSpec) string {
+	csr, err := ParseCSR(&certificatesv1beta1.CertificateSigningRequest{Spec: *obj})
```
suggest changing this helper to just need the `obj.Spec.Request` field so you don't need to artifically construct this

liggitt commented on staging/src/k8s.io/kubectl/pkg/describe/versioned/describe.go:
```diff
@@ -3247,6 +3247,9 @@ func describeCertificateSigningRequest(csr *certificatesv1beta1.CertificateSigni
 		w.Write(LEVEL_0, "CreationTimestamp:\t%s\n", csr.CreationTimestamp.Time.Format(time.RFC1123Z))
 		w.Write(LEVEL_0, "Requesting User:\t%s\n", csr.Spec.Username)
+		if csr.Spec.SignerName != nil {
+			w.Write(LEVEL_0, "Signer:\t%s\n", *csr.Spec.SignerName)
```
also add to pkg/printers/internalversion/printers.go as a new column:
```
	certificateSigningRequestColumnDefinitions := []metav1.TableColumnDefinition{
		{Name: "Name", Type: "string", Format: "name", Description: metav1.ObjectMeta{}.SwaggerDoc()["name"]},
		{Name: "Age", Type: "string", Description: metav1.ObjectMeta{}.SwaggerDoc()["creationTimestamp"]},
		{Name: "Requestor", Type: "string", Description: certificatesv1beta1.CertificateSigningRequestSpec{}.SwaggerDoc()["request"]},
		{Name: "Condition", Type: "string", Description: certificatesv1beta1.CertificateSigningRequestStatus{}.SwaggerDoc()["conditions"]},
	}
	h.TableHandler(certificateSigningRequestColumnDefinitions, printCertificateSigningRequest)
	h.TableHandler(certificateSigningRequestColumnDefinitions, printCertificateSigningRequestList)
```

munnerz replied:
:+1: although worth noting that the internal version does not use a `*string` anymore, so the `if` needs to check for `""` instead (which I think is okay)

liggitt commented:
Feel free to squash to a single commit when addressing comments (as far as I'm concerned)

munnerz commented:
@liggitt addressed your feedback - I squashed this down to two commits, one containing your review changes.

Let me know if you want me to squash further or want any more changes 😄

liggitt commented on pkg/apis/certificates/validation/validation.go:
```diff
@@ -48,19 +51,95 @@ func ValidateCertificateRequestName(name string, prefix bool) []string {
+
+	segments := strings.Split(signerName, "/")
+	// validate that there is one '/' in the signerName.
+	// we do this after validating the domain segment to provide more info to the user.
```
nit: this comment is no longer accurate

liggitt commented:
unit tests need updating

liggitt commented:
as do generated files

liggitt commented:
lgtm, update tests and generated files and squash

we'll hold this until the controller manager implementation is ready on top of this commit

enj commented on pkg/printers/internalversion/printers.go:
```diff
@@ -1713,7 +1714,11 @@ func printCertificateSigningRequest(obj *certificates.CertificateSigningRequest,
 	}
-	row.Cells = append(row.Cells, obj.Name, translateTimestampSince(obj.CreationTimestamp), obj.Spec.Username, status)
+	signerName := "<none>"
+	if obj.Spec.SignerName == "" {
```
You have this backwards:

`if len(obj.Spec.SignerName) != 0 {`

liggitt replied:
yeah, also check for nil:
```
if obj.Spec.SignerName != nil && len(obj.Spec.SignerName) > 0 {
```

munnerz replied:
Fixed, added a test for this case and fixed the failing tests 😄

enj reviewed (approved):
I just did a quick pass under the assumption that you had not changed this much.

liggitt commented:
go ahead and squash this to a single commit to cleanly layer the other PRs on top of it

liggitt commented:
/retest

liggitt commented:
/lgtm
/approve
/hold for admission and controller implementation PRs

munnerz commented:
Changed due to rebasing on master, as #86933 which depends on this had hit a conflict.

enj commented:
/lgtm

(just retroactive assertion that this is good)
